gcloud-java-dns/src/main/java/com/google/gcloud/testing/LocalDnsHelper.java

```
@@ -0,0 +1,1426 @@
+    Change result = OptionParsersAndExtractors.extractFields(completeChange, fields);
+    try {
+      return new Response(HTTP_OK, jsonFactory.toString(result));
+    } catch (IOException e) {
+      return Error.INTERNAL_ERROR.response(
+          String.format("Error when serializing change %s in managed zone %s in project %s.",
+              result.getId(), zoneName, projectId));
+    }
+  }
+
+  /**
+   * Applies change. Uses a new thread which applies the change only if DELAY_CHANGE is > 0.
+   */
+  private Thread invokeChange(final String projectId, final String zoneName,
+      final String changeId) {
+    if (delayChange > 0) {
+      Thread thread = new Thread() {
+        @Override
+        public void run() {
+          try {
+            Thread.sleep(delayChange); // simulate delayed execution
+          } catch (InterruptedException ex) {
+            log.log(Level.WARNING, "Thread was interrupted while sleeping.", ex);
+          }
+          // start applying the changes
+          applyExistingChange(projectId, zoneName, changeId);
+        }
+      };
+      thread.start();
+      return thread;
+    } else {
+      applyExistingChange(projectId, zoneName, changeId);
+      return null;
+    }
+  }
+
+  /**
+   * Applies changes to a zone. Repeatedly tries until succeeds. Thread safe and deadlock safe.
+   */
+  private void applyExistingChange(String projectId, String zoneName, String changeId) {
+    Change change = findChange(projectId, zoneName, changeId);
+    if (change == null) {
+      return; // no such change exists, nothing to do
+    }
+    ZoneContainer wrapper = findZone(projectId, zoneName);
+    ConcurrentSkipListMap<String, ImmutableList<RrsetWrapper>> dnsRecords = wrapper.dnsRecords();
+    while (true) {
+      // managed zone must have a set of records which is not null
+      ImmutableList<RrsetWrapper> original = dnsRecords.get(zoneName);
+      assert original != null;
```

This assert failing would cause the server to shutdown quietly, which isn't ideal.  I think if this condition isn't met, the server should return an internal error with an error message.